server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -231,6 +231,19 @@ public boolean allPrimaryShardsActive() {
         return primaryShardsActive() == shards.length;
     }
 
+    /**
+     * @return <code>true</code> if all search shards are active. This is useful to determine if an index is available to service queries.
+     */
+    public boolean allSearchShardsActive() {
+        int counter = 0;
+        for (IndexShardRoutingTable shardRoutingTable : this.shards) {
+            if (shardRoutingTable.allShards().anyMatch(s -> s.active() && s.role().isSearchable())) {
```

Comment:
Does method require _every_ search shard copy to be active? It looks like `anyMatch` would return true even if only one of multiple shard copies of the same shard is active.

Additionally it might be good to avoid stream from within the loop

Reply:
> It looks like anyMatch would return true even if only one of multiple shard copies of the same shard is active.

That's the intention: this is to replace `allPrimaryShardsActive()` as a way to determine if an index is ready to search. But the name is confusing, I also interpret it as returning whether _all_ the search shard copies are active.

Can we call it something like `allShardsSearchReady` instead?

Also +1 to avoiding streams here, and also no need for `counter`, we can just `return false` as soon as we find any shard which isn't search-ready.

Reply:
++ I would agree, this method is pretty suboptimal performance wise. It's fine if it's run like this for a limited set of indices but if it's executed with a wider scope we should either optimise it to avoid the stream usage or allocation through `org.elasticsearch.cluster.routing.IndexShardRoutingTable#shard` or move the code to where it's used for system indices only(right?).

Reply:
Also, I would like to note we should use `org.elasticsearch.cluster.routing.OperationRouting#canSearchShard()` which takes into account the new fast refresh setting in Serverless. And I would suggest to extend the test to have a fast refresh index. Maybe it's a good idea to also put a comment above the isSearchable() method that people should ideally use the canSearchShard() method, just to remind.

Reply:
> But the name is confusing,

I see that now...sorry for the confusion. How about `readyForSearch` ? `all` is still a bit ambiguous w.r.t. every vs. needed

>  Additionally it might be good to avoid stream from within the loop / this method is pretty suboptimal performance wise / Also +1 to avoiding streams here

It is difficult to know that up front. The loop here will generally be small and it difficult to know how much of a performance hit using stream API actually is. Is it the stream allocation, it is the number of iteratation within the stream, how do we know that anyMatch from a stream is not more performant than manual iteration.  Additionally, the API only exposes this via a Stream.of (which is a pretty recent addition). I assume it is exposed as stream to avoid exposing the raw array. So to avoid the stream I would need to expose the raw array and to do that i would need to make an immutable copy ... but that seems as expensive as using a Stream. This code might be executed in a way that is extremely sensitive to performance (like frequent cluster state updates) but it is not today and don't think it's usage in that manner is likely (maybe it is?). I can probably find a way to avoid the stream or the copy but I question if that needed and/or if it is a premature optimization. 

> or move the code to where it's used for system indices only(right?).

I can move the code to only service my usecase(s) but thought this would be generally useful. 

> and also no need for counter, we can just return false as soon as we find any shard which isn't search-ready.

++

> Also, I would like to note we should use org.elasticsearch.cluster.routing.OperationRouting#canSearchShard() which takes into account the new fast refresh setting. 

Thanks for calling that out. I had not really considered fast refresh. I had assumed that since fast refresh indices don't have replicas the primary would be marked as both search and index which works fine (and is tested) with isSearchable().  Can you help me understand why that is not the case ?

Reply:
> how do we know that anyMatch from a stream is not more performant than manual iteration. 

lots of benchmarking for many-shards and such + the theory here is quite clear as well. Unless the whole thing inlines (which it won't here for standard JVM settings) you have far more code and virtual method calls to execute with the stream compared to a loop + have allocations that burn more cycles outside the main execution path.

> I can probably find a way to avoid the stream or the copy but I question if that needed and/or if it is a premature optimization.

Why would you need any copy here? You can just iterate like we do elsewhere via:

```java
    for (IndexShardRoutingTable shardRoutingTable : this.shards) {
            for (int idx = 0; copy < shardRoutingTable.size(); idx++) {
```

can't you?

>

Reply:
The copy comment was in reference to the array but I see now that we expose access to a single entry from the array. I will use that.

I believe and trust your assessment. The push back here is more about without your (and others in thread) assessment a) knowing when/where we should avoid streams b) the existence of APIs that expose the stream we are not supposed to use c) knowing the performance characteristics for alternatives to streams. 

I understand the motivation and need to keep things fast, but I am failing to know how to reason about this level of optimization in practice. When does the expected usage / area of code elevate itself to the point when I need critically think about how the code will be inlined and virtual method calls ? Do these optimization concerns extend beyond Stream usage ? Should there be warning in the code for these areas ? Short of running micro bench marks (or asking you :)) how do I know that the alternative is actually more performant and not a incorrect micro optimization?

Reply:
> Can you help me understand why that is not the case ?

Unfortunately the `isSearchable()` function is depending on the shard's role which is an enum. While to know whether an index is a fast refresh index, we need to access the index metadata. So we created a new function canSearchShard that accesses both arguments to make the final judgement.